core/lib/dal/src/connection/mod.rs

```diff
@@ -170,24 +162,68 @@ impl TestTemplate {
             .await
             .context("CREATE DATABASE")?;
 
-        const TEST_MAX_CONNECTIONS: u32 = 50; // Expected to be enough for any unit test.
-        ConnectionPool::builder(self.url(&db_new).as_ref(), TEST_MAX_CONNECTIONS)
-            .build()
-            .await
-            .context("ConnectionPool::builder()")
+        Ok(ConnectionPool::builder(
+            self.url(&db_new).as_ref(),
+            connections,
+        ))
+    }
+}
+
+#[derive(Clone)]
+pub struct ConnectionPool {
+    pub(crate) inner: PgPool,
+    database_url: String,
+    max_size: u32,
+    traced_connections: Option<Arc<TracedConnections>>,
+}
+
+impl fmt::Debug for ConnectionPool {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // We don't print the `database_url`, as is may contain
+        // sensitive information (e.g. database password).
+        formatter
+            .debug_struct("ConnectionPool")
+            .field("max_size", &self.max_size)
+            .finish_non_exhaustive()
     }
 }
 
 impl ConnectionPool {
+    const TEST_ACQUIRE_TIMEOUT: Duration = Duration::from_secs(1);
+
+    /// Creates a test pool with a reasonably large number of connections.
+    ///
+    /// Test pools trace their active connections. If acquiring a connection fails (e.g., with a timeout),
+    /// the returned error will contain information on all active connections.
     pub async fn test_pool() -> ConnectionPool {
-        TestTemplate::empty().unwrap().create_db().await.unwrap()
+        const DEFAULT_CONNECTIONS: u32 = 50; // Expected to be enough for any unit test.
+        Self::constrained_test_pool(DEFAULT_CONNECTIONS).await
+    }
+
+    /// Same as [`Self::test_pool()`], but with a configurable number of connections. This is useful to test
+    /// behavior of components that rely on singleton / constrained pools in production.
```

Comment:
It may be my problem, but I've had issues with the state keeper logic a couple of times where it passed unit tests, but hanged up on integration tests / loadtest because of a connection acquisition deadlock (the state keeper is provided with a singleton pool). Now it's possible to test this using unit tests and pinpoint where the extra connection has been created.